File: Search in Rotated Sorted Array II/Search in Rotated Sorted Array II.java
public class Solution {
    public boolean search(int[] A, int target) {
        int head = 0;
        int end = A.length - 1;
        
        while(A[head] == A[end]){
            --end;
            if(end < 0){
                return target == A[0];
            }
        }
        
        while(head <= end){
            int mid = head + end;
            mid /= 2;
            
            if(target == A[mid])        return true;
            else if(target < A[mid]){
                if((A[head] < A[mid] && target >= A[head]) || (A[head] > A[mid] && target < A[head])){
                    end = mid - 1;
                }
                else{
                    head = mid + 1;
                }
            }
            else{   // target > A[mid]
                if((A[mid] < A[end] && target <= A[end]) || (A[mid] > A[end] && target > A[end])){
                    head = mid + 1;
                }
                else{
                    end = mid - 1;
                }
            }
        }
        return false;
    }
}

File: Search in Rotated Sorted Array II/Search in Rotated Sorted Array II.cpp
class Solution {
public:
    bool search(int A[], int n, int target) {
        int l = 0;
        int r = n - 1;
        
        while(l <= r){
            int m = (l + r) / 2;
            //if((A[l] == target) || (A[m] == target) || (A[r] == target))      return true;
            if(A[m] == target)      return true;
            
            if(A[l] < A[m]){
                if((target >= A[l]) && (target < A[m])){        // Choose left half
                    r = m - 1;
                }
                else{
                    l = m + 1;
                }
            }
            else if(A[l] > A[m]){       // A[l] > A[m] indicates pivot in the left half
                if((target > A[m]) && (target <= A[r])){
                    l = m + 1;
                }
                else{
                    r = m - 1;
                }
            }
            else{
                ++l;
            }
        }
        
        return false;
    }
};
